src/metabase/driver/util.clj

```diff
@@ -215,18 +215,19 @@
    critical metabase features that use this check."
   5000)
 
-(defn supports?
-  "A defensive wrapper around [[database-supports?]]. It adds logging and error handling to avoid crashing the app if this
-   method takes a long time to execute or throws an exception. This is useful because `supports?` is used in so many critical
-   places in the app, and we don't want a single driver to crash the app if it throws an exception, or delay the user if it
-   takes a long time to execute."
-  [driver feature database]
-  (try
-    (u/with-timeout supports?-timeout-ms
-      (driver/database-supports? driver feature database))
-    (catch Throwable e
-      (log/error e (u/format-color 'red "Failed to check feature '%s' for database '%s'" (name feature) (:name database)))
-      false)))
+(def ^{:arglists '([driver feature database])} supports?
+  "A defensive wrapper around [[database-supports?]]. It adds logging, caching, and error handling to avoid crashing the app
+   if this method takes a long time to execute or throws an exception. This is useful because `supports?` is used in so many
+   critical places in the app, and we don't want a single driver to crash the app if it throws an exception, or delay the user
+   if it takes a long time to execute."
+  (mdb/memoize-for-application-db ; memoize because this can be called in a tight loop, and will only change if the database changes versions
```

Comment:
Had to read this several times—it memoizes for the unique pair of {data DB, app DB}.